File: r3.py
#!/usr/bin/env python

# This program uses itm.py to solve the problem presented in section 7.3 of 
# "A Guide to the Use of the ITS Irregular Terrain Model in the Area
# Prediction Mode," Hufford, Longley, Kissick, NTIA Report 82-100, 1982.
#
# Except for main() at bottom, subroutines are in alphabetical order.
#
# Mike Markowski, mike.ab3ap@gmail.com
# June 2023

from scipy.stats import norm
import itm
import matplotlib.pyplot as plt
import numpy as np

def predict():
    '''Implement problem 7.2 Optimum Television Station Separation, p. 42 in 
    "A Guide to the Use of the ITS Irregular Terrain Model in the Area
    Prediction Mode," Hufford, Longley, Kissick, NTIA Report 82-100, 1982.
    '''

    # Set up values for R3 data set as given in section 7.3 of report.
    itm.dictionaries()       # Prepare ITM.
    f_MHz = itm.prop['f_MHz'] = 410 # UHF tv-like.
    itm.prop['hg'][0] = 275  # m, transmit antenna structure height.
    itm.prop['hg'][1] = 6.6  # m, receive antenna structure height.
    itm.prop['dh'] = 126     # m, roughness factor for hilly terrain.
    itm.prop['klim'] = itm.cContTemp # Continental temperate.
    itm.mdvarSet(itm.mBcast) # Broadcast mode.
    pol = 'h'     # Horizontal antenna polarization.
    epsilon = 15  # F/m, relative ground permittivity.  Average ground.
    sigma = 0.005 # S/m, relative ground conductivity.  Average ground.
    zSys = 1700   # m, system elevation above sea level.
    N0 = 300      # N-units, sea level refractivity (yields Ns=250).
    zT = 0        # qT = 0.5, no time variability.

    itm.lrPrep(f_MHz, zSys, N0, pol, epsilon, sigma)
    itm.lrArea(['r', 'r']) # Recreate figure 9.
#    itm.lrArea(['v', 'v']) # Recreate figure 12.

    plotsX = [] # Start collection of curves.
    plotsY = []
    labels = []
    # All curves exactly match report except A(0.1,0.1).
    for qL,qC in [[0.9,0.9],[0.5,0.9],[0.5,0.5],[0.5,0.1],[0.1,0.1]]:
        zL = norm.isf(qL) # Quantile of location.
        zC = norm.isf(qC) # Quantile of confidence (situation).
        plotX = [] # Start new curve.
        plotY = []
        for dist_km in range(1, 101): # Distance from cliff-top transmitter.
            itm.lrProp(dist_km*1e3) # Longley-Rice propagation calculation.
            aRef_dB,_ = itm.aVar(zT, zL, zC) # Variability of attenuation.
            plotX.append(dist_km)
            plotY.append(aRef_dB)
            if aRef_dB > 40: # Match max atten in Figures 9 and 12.
                break
        plotsX.append(plotX) # Save new curve.
        plotsY.append(plotY)
        labels.append('A(%.1f, %.1f)' % (qL, qC)) # This curve's label.
    return plotsX, plotsY, labels

def present(x, y, lbl):
    # Graph settings to match Figures in report.
    fig = plt.figure(figsize=(8, 4), dpi=120)
    ax = fig.add_subplot(1, 1, 1)
    ax.set_title('Predicted Attenuation, R3 Data')
    ax.set_xlabel('Distance (km)')
    ax.set_ylabel('Attenuation (dB)')
    ax.set_xlim(0, 100)
    ax.set_ylim(-9.9, 40)
    ax.invert_yaxis()
    ax.grid(True)

    for i in range(len(x)):
        plt.plot(x[i], y[i], lw=0.75, label=lbl[i])
    plt.legend()
    plt.show()

def main():
    x, y, lbl = predict()
    present(x, y, lbl)

if __name__ == '__main__':
    main()

File: r3.f
c     This program uses itm.for to solve the problem presented in section 7.3
c     of "A Guide to the Use of the ITS Irregular Terrain Model in the Area
c     Prediction Mode," Hufford, Longley, Kissick, NTIA Report 82-100, 1982.
c
c     Mike Markowski, mike.ab3ap@gmail.com
c     June 2023

      subroutine predict()

      common/prop/kwx,aref,mdp,dist,hg(2),wn,dh,ens,gme,zgnd,
     x   he(2),dl(2),the(2)
         complex zgnd
      common/propv/lvar,sgc,mdvar,klim
      common/propa/dlsa,dx,ael,ak1,ak2,aed,emd,aes,ems,dls(2),dla,tha
      dimension kst(2)
      integer dist_km

c     Implement problem 7.2 Optimum Television Station Separation, p. 42 in 
c     "A Guide to the Use of the ITS Irregular Terrain Model in the Area
c     Prediction Mode," Hufford, Longley, Kissick, NTIA Report 82-100, 1982.

c     Set up values for R3 data set as given in section 7.3 of report.
c     prop
      dh = 126.    ! m, roughness factor for hilly terrain.
      hg(1) = 275. ! m, transmit antenna structure height.
      hg(2) = 6.6  ! m, receive antenna structure height.

c     propv
      lvar = 5
      mdvar = 3    ! Broadcast mode.

      en0 = 300.   ! N-units, sea level refractivity (yields Ns=250).
      eps = 15.    ! F/m, relative ground permittivity.  Average ground.
      f_MHz = 410. ! UHF tv-like.
      ipol = 0     ! Horizontal antenna polarization.
      sgm = 0.005  ! S/m, relative ground conductivity.  Average ground.
      zs = 1700.   ! m, system elevation above sea level.
      zT = 0.      ! qT = 0.5, no time variability.

      kwx = 0
      call qlrps(f_MHz, zs, en0, ipol, eps, sgm)
      if (.true.) then ! Recreate figure 9.
        kst(1) = 0
        kst(2) = 0
      else ! Recreate figure 12.
        kst(1) = 2
        kst(2) = 2
      endif
      call qlra(kst, -1, -1)

c     All curves exactly match report except A(0.1,0.1).
      do i=1,5
        if (i .eq. 1) then
          qL = 0.9
          qC = 0.9
        elseif (i .eq. 2) then
          qL = 0.5
          qC = 0.9
        elseif (i .eq. 3) then
          qL = 0.5
          qC = 0.5
        elseif (i .eq. 4) then
          qL = 0.5
          qC = 0.1
        elseif (i .eq. 5) then
          qL = 0.1
          qC = 0.1
        endif
 100    format(2hA(f4.1, f4.1,1h))
        write(*, 100) qL, qC
        zL = qerfi(qL)   ! Quantile of location.
        zC = qerfi(qC)   ! Quantile of confidence (situation).
        do dist_km=1,100 ! Distance from cliff-top transmitter.
          lvar = max(1, lvar)
          call lrProp(dist_km*1e3)   ! Longley-Rice propagation calculation.
          aRef_dB = aVar(zT, zL, zC) ! Variability of attenuation.
c200      format(i6, f8.1)
c         write(*, 200) dist_km, aRef_dB
 200      format(f8.1,1x ,$)
          write(*, 200) aRef_dB
          if (aRef_dB .gt. 40) then  ! Match max atten in Figures 9 and 12.
            exit
          endif
        enddo
        write(*,*)
      enddo
      endsubroutine

      call predict()
      end
